### PR TITLE
Adding ORBIT-2 global precip downscaling module to earth2studio

### PR DESCRIPTION
**Pull request by irlyngaas**
# Earth2Studio Pull Request

## Description

Adding the ORBIT-2 module to e2s. Current module includes a model for downscaling precipitation 

None

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/earth2studio/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/earth2studio/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/earth2studio/issues) is linked to this pull request.
- [ ] Assess and address Greptile feedback (AI code review bot for guidance; use discretion, addressing all feedback is not required).

## Dependencies
ORBIT-2, which is added in the pyproject.toml

### CODE REVIEW

File: earth2studio/run_orbit.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    t2_max = torch.max(t2_sst_combined, dim=1).values
+    t2_min = torch.min(t2_sst_combined, dim=1).values
+    total_p_24hr = total_p_24hr.unsqueeze(1)
+    t2_max = t2_max.unsqueeze(1)
+    t2_min = t2_min.unsqueeze(1)
+    x = torch.cat((x,total_p_24hr,t2_max,t2_min), dim=1)
+    logger.success(f"Fetched data from {data.__class__.__name__}")
+
+    if data_check:
+        check_data(x[0].unsqueeze(0), time[0], orbit.in_variables)
```
[the remark this answers is not in the document]
Comment (irlyngaas) [reply earth2studio/run_orbit.py]: This can be deleted

File: earth2studio/models/dx/orbit_2_precip.py [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+        land_sea_mask = np.load(package.resolve("static_variables/land_sea_mask_0.25deg.npy"))
+        orography = np.load(package.resolve("static_variables/orography_0.25deg.npy"))
+        lattitude = np.load(package.resolve("static_variables/lattitude_0.25deg.npy"))
+        landcover = np.load(package.resolve("static_variables/landcover_0.25deg.npy"))
+
+        normalize_mean_lowres = np.load(package.resolve("mean_std/era5/0.25_deg/normalize_mean.npz"))
+        normalize_std_lowres = np.load(package.resolve("mean_std/era5/0.25_deg/normalize_std.npz"))
+
+        normalize_mean_highres = np.load(package.resolve("mean_std/era5/0.25_deg/normalize_mean.npz"))
+        normalize_std_highres = np.load(package.resolve("mean_std/era5/0.25_deg/normalize_std.npz"))
```
[the remark this answers is not in the document]
Comment (irlyngaas) [reply earth2studio/models/dx/orbit_2_precip.py]: This is correct based on how this model and input data is setup. We don't have 0.0625 deg era5 data so we can't get normalization statistics for it. Thus we just use the statistics from the 0.25 deg data to do it

File: earth2studio/models/dx/orbit_2_precip.py [hunk 3]
```diff
@@ -0,0 +1,554 @@
...
+            lon=np.arange(0, 360, 0.25)
+        )
+
+        # Build regridder (stays on GPU)
+        regridder = earth2grid.get_regridder(src_grid, dst_grid)
+        regridder = regridder.to(device).float()
+
+        # Regrid — expects (..., lat, lon), handles time+variable dims automatically
+        # x shape: (time, variables, lat, lon)
+        x = regridder(x)
```
[the remark this answers is not in the document]
Comment (irlyngaas) [reply earth2studio/models/dx/orbit_2_precip.py]: I'm open to suggestions on how you all would implement this

File: earth2studio/run_orbit.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        if k != "batch" # remove placeholder batch dim
+    })
+    total_coords["time"] = time
+    total_coords.move_to_end("time", last=False)
+
+    io.add_array(total_coords, total_coords["variable"], overwrite=True)
+    logger.info("Inference Starting")
+    x, output_coords = orbit(x, input_coords)
+    io.write(x, output_coords, total_coords["variable"][0])
+    logger.success("Inference complete")
```
[the remark this answers is not in the document]
Comment (irlyngaas) [reply earth2studio/run_orbit.py]: Give me a suggested return value

File: launch.sh [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+module load anaconda3
+conda activate orbit-e2s
+
+export HOSTNAME=$(hostname)
+export PYTHONNOUSERSITE=1
+
+export OMP_NUM_THREADS=7
+time srun -n $((SLURM_JOB_NUM_NODES*1)) \
+python ./launch_test.py
+
```
[the remark this answers is not in the document]
Comment (irlyngaas) [reply launch.sh]: This launch script will be deleted

File: earth2studio/run_orbit.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        k: v for k, v in
+        orbit.output_coords(orbit.input_coords()).items()
+        if k != "batch" # remove placeholder batch dim
+    })
+    total_coords["time"] = time
+    total_coords.move_to_end("time", last=False)
+
+    io.add_array(total_coords, total_coords["variable"], overwrite=True)
+    logger.info("Inference Starting")
+    x, output_coords = orbit(x, input_coords)
```
[the remark this answers is not in the document]
Comment (irlyngaas) [reply earth2studio/run_orbit.py]: I've accounted for this. Some messy things needed to be done get the data into the shape that works with ORBIT, including renaming variables. Everything looks correct to me.

File: earth2studio/data/ncar.py [hunk 7]
```diff
@@ -44,6 +44,8 @@
 logger.remove()
 logger.add(lambda msg: tqdm.write(msg, end=""), colorize=True)
 
+import pandas as pd
+
 
 @dataclass
```
[the remark this answers is not in the document]
Comment (irlyngaas) [reply earth2studio/data/ncar.py:47]: I removed this

File: earth2studio/models/dx/orbit_2_precip.py [hunk 8]
```diff
@@ -0,0 +1,554 @@
...
+
+        map_location = "cpu"
+        checkpoint = torch.load(package.resolve(model_type+"-finetune"+"/"+model_type+"_"+model_size+"_"+model_variable+".ckpt"), map_location=map_location)
+
+        pretrain_model = checkpoint["model_state_dict"]
+        del checkpoint
+
+        state_dict = model.state_dict()
+
+        model.load_state_dict(pretrain_model)
```
[the remark this answers is not in the document]
Comment (irlyngaas) [reply earth2studio/models/dx/orbit_2_precip.py]: I updated this

File: earth2studio/models/dx/__init__.py [hunk 9]
```diff
@@ -55,4 +57,5 @@
     "SolarRadiationAFNO1H",
     "SolarRadiationAFNO6H",
     "WindgustAFNO",
+    "OrbitGlobalPrecip9_5M",
 ]
```
[the remark this answers is not in the document]
Comment (irlyngaas) [reply earth2studio/models/dx/__init__.py:60]: Fixed

File: earth2studio/run_orbit.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (irlyngaas) [reply earth2studio/run_orbit.py]: Added

File: earth2studio/models/dx/orbit_2_precip.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (irlyngaas) [reply earth2studio/models/dx/orbit_2_precip.py]: Added fixes for this

**Comment by NickGeneva** [comment]
Thanks for opening this @irlyngaas 

Any chance you can add some unit tests for this similar to other DX models? For example: 
https://github.com/NVIDIA/earth2studio/blob/main/test/models/dx/test_precip_afno_v2.py

They are really helpful to also help understand running the model!
My quick look at the wrapper, it seems good. Will check in more detail shortly.

**Comment by NickGeneva** [comment]
@greptile-apps

**Comment by irlyngaas** [comment]
Ok I went ahead and added all of those unit tests.

**Comment by NickGeneva** [comment]
Summary of clean up / edits:

- 20 input variables — 15 atmospheric, swvl1, sst, tp24, t2m_max, t2m_min; SST is consumed during preprocessing (merged with t2m over ocean), leaving 19 dynamic + 4 static = 23 internal channels
- Precipitation in log-space — input precip normalized via log1p(mm) with 0.25 mm/day threshold; output denormalized via expm1 then mm→m conversion
- Normalization as registered buffers — all mean/std tensors and static fields (land_sea_mask, orography, lattitude, landcover) stored as register_buffer for automatic device placement
- Latitude convention handling — E2S uses descending 90→-90; ORBIT uses ascending -90→90; model flips internally and flips output back
- Optional dependency pattern — climate-learn wrapped in try/except ImportError with OptionalDependencyFailure
- Weights from HuggingFace — pinned commit from jychoi-hpc/ORBIT-2
- No earth2grid regridder — removed in favor of manual lat flip + pole row removal

Sample plotting script, looks good at least to my eyes!

<details>

```python
"""Plot ORBIT-2 predicted precipitation alongside ERA5 input tp24.

Side-by-side comparison: ERA5 tp24 (coarse) | ORBIT-2 output (downscaled).
"""

import numpy as np
import torch
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
import cartopy.feature as cfeature
from collections import OrderedDict
from loguru import logger

from earth2studio.data import NCAR_ERA5, prep_data_array
from earth2studio.models.dx import OrbitGlobalPrecip
from earth2studio.utils.time import to_time_array

# ============================================================================
# Configuration
# ============================================================================
TIME = [np.datetime64("2023-06-15", "ns")]
logger.info(f"Date: {TIME[0]}")

# ============================================================================
# Load model and data source
# ============================================================================
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
logger.info(f"Device: {device}")

package = OrbitGlobalPrecip.load_default_package()
orbit = OrbitGlobalPrecip.load_model(package, model_size="126m")
orbit = orbit.to(device)

data = NCAR_ERA5()
time = to_time_array(TIME)

# ============================================================================
# Fetch base variables (without tp24, t2m_max, t2m_min)
# ============================================================================
x, coords = prep_data_array(
    data(time, orbit.input_coords()["variable"][:-3]), device=device
)

# ============================================================================
# Build derived inputs: tp24, t2m_max, t2m_min from 24 hourly steps
# ============================================================================
for i in range(24):
    time_ = np.array(time) + np.timedelta64(1, "h") * (-1 * i)
    p, _ = prep_data_array(data(time_, ["cp", "lsp", "t2m", "sst"]), device=device)

    batch_split = torch.split(p, 1)
    if i == 0:
        batch_p = torch.zeros(
            (len(time), 24, p.shape[-3], p.shape[-2], p.shape[-1]), device=device
        )
    for j in range(len(batch_split)):
        batch_p[j, i] = batch_split[j][0]

total_p_24hr = batch_p[:, :, 0] + batch_p[:, :, 1]
total_p_24hr = torch.sum(total_p_24hr, dim=1)
t2_sst_combined = torch.where(
    torch.isnan(batch_p[:, :, 3]), batch_p[:, :, 2], batch_p[:, :, 3]
)
t2_max = torch.max(t2_sst_combined, dim=1).values
t2_min = torch.min(t2_sst_combined, dim=1).values

# Keep ERA5 tp24 for plotting (convert m -> mm)
era5_tp24 = np.sum(batch_p[0, :, 1].cpu().numpy() * 1000.0, axis=0)

total_p_24hr = total_p_24hr.unsqueeze(1)
t2_max = t2_max.unsqueeze(1)
t2_min = t2_min.unsqueeze(1)
x = torch.cat((x, total_p_24hr, t2_max, t2_min), dim=1)
logger.success("Input data assembled")

# ============================================================================
# Run ORBIT-2 inference
# ============================================================================
input_coords = OrderedDict(
    {k: v for k, v in orbit.input_coords().items() if k != "batch"}
)
input_coords["time"] = time
input_coords.move_to_end("time", last=False)

logger.info("Running ORBIT-2 inference...")
output, output_coords = orbit(x, input_coords)
logger.success("Inference complete")

# Output in meters, convert to mm
orbit_precip = output[0, 0].cpu().numpy() * 1000.0

# ============================================================================
# Plot side by side
# ============================================================================
fig, axes = plt.subplots(
    1, 2, figsize=(16, 5),
    subplot_kw={"projection": ccrs.PlateCarree()},
)

# ERA5 input tp24 (0.25 deg)
era5_lat = orbit.input_coords()["lat"]
era5_lon = orbit.input_coords()["lon"]
# tp24 is computed after flip+trim in preprocess, but here we have it pre-flip
# ERA5 lat is [90, -90] with 721 points; our era5_tp24 is (721, 1440)
vmax = max(np.nanpercentile(era5_tp24, 99.5), np.nanpercentile(orbit_precip, 99.5))

im0 = axes[0].pcolormesh(
    era5_lon, era5_lat, era5_tp24, cmap="viridis", vmin=0, vmax=vmax,
    transform=ccrs.PlateCarree(),
)
axes[0].add_feature(cfeature.COASTLINE, linewidth=0.5)
axes[0].add_feature(cfeature.BORDERS, linewidth=0.3)
axes[0].set_title("ERA5 tp24 (0.25°)")
plt.colorbar(im0, ax=axes[0], fraction=0.046)

# ORBIT-2 output (0.0625 deg)
orbit_lat = output_coords["lat"]
orbit_lon = output_coords["lon"]

im1 = axes[1].pcolormesh(
    orbit_lon, orbit_lat, orbit_precip, cmap="viridis", vmin=0, vmax=vmax,
    transform=ccrs.PlateCarree(),
)
axes[1].add_feature(cfeature.COASTLINE, linewidth=0.5)
axes[1].add_feature(cfeature.BORDERS, linewidth=0.3)
axes[1].set_title("ORBIT-2 Downscaled (0.0625°)")
plt.colorbar(im1, ax=axes[1], fraction=0.046)

plt.suptitle(f"24h Precipitation (mm) — {TIME[0]}", fontsize=14)
plt.tight_layout(rect=[0, 0, 1, 0.95])
plt.savefig("orbit2_vs_era5.png", dpi=150, bbox_inches="tight")
logger.info("Saved plot to orbit2_vs_era5.png")
plt.show()

```

</details>

orbit2_vs_era5

**Comment by NickGeneva** [comment]
@greptile-apps

**Comment by NickGeneva** [comment]
coverage:

```
earth2studio/models/dx/orbit2_precip.py             235     16    93%   286-287, 392, 409-414, 433, 439-440, 707-712
```

File: earth2studio/models/dx/orbit2_precip.py [hunk 10]
```diff
@@ -0,0 +1,728 @@
...
+        """Adjust coordinates after vertical image flip."""
+        # Adjust input tile extraction coordinates after flip
+        # Calculate flipped y-coordinates for tile extraction
+        tile_height_with_overlap = processor.yinp // processor.div + (
+            processor.top + processor.bottom
+        )
+        yi2tp = tile_height_with_overlap - coords.yi1t
+        yi1tp = tile_height_with_overlap - coords.yi2t
+        coords.yi1t = yi1tp
+        coords.yi2t = yi2tp
```
[the remark this answers is not in the document]
Comment (NickGeneva) [reply earth2studio/models/dx/orbit2_precip.py:614]: Does actually matter for thsi model since precip is log normalized. But fixed anyway

**Comment by NickGeneva** [comment]
@greptile-apps

File: earth2studio/models/dx/orbit2_precip.py [hunk 11]
```diff
@@ -0,0 +1,730 @@
...
+                    tiles.append(yhat)
+            yhat = self.stitch_tiles(tiles, tile_coords, processor, yhat.shape[0])
+        else:
+            x = self.preprocess_input(x)
+            yhat = self.model.forward(
+                x, self.core_input_variables, self.core_output_varaibles
+            )
+            yhat = self.clip_replace_constant(yhat, self.core_output_varaibles)
+            yhat = self._denormalize_output(yhat)
+        return yhat
```
[the remark this answers is not in the document]
Comment (NickGeneva) [reply earth2studio/models/dx/orbit2_precip.py:716]: Fixed.

**Comment by NickGeneva** [comment]
@greptile-apps

**Review by NickGeneva — APPROVED** [review]
Merging, can follow up in other PRs. Want to get the initial model added since results look good.
